### PR TITLE
UART HIL redesign with asynchronous calls

### PR DESCRIPTION
**Pull request by brghena**
UART updates
- Updates UART HIL to an asynchronous interface
- Updates USART implementation of UART to fit and to generally be better
- Updates Console and nrf-serialization to fit
- Updates `io.rs` with new UART interface

DMA updates
- DMA peripheral stuff will mostly disappear once PR #125 hits (intend to merge this after #125)
- Added all the possible DMA interrupts so no one would ever forget to do so again

`take_cell` updates
- added `map_or` function that had a good use as a common idiom in UART and SPI to reduce possible mistakes. Otherwise, implementation requires a `take()` and then remembering to `replace()` in the same code block

Missing Functionality
- `io.rs` currently requires some backdoor methods into the USART until we can figure out how to send static strings to the UART
- automatically disabling tx peripheral is not performed. Turns out that even once the DMA transfer is "finished", the UART is still sending. Need to use interrupts to figure out when transmission is actually complete in order to disable the peripherals
- assumes 48 MHz operation. We need an interface for peripherals to determine the core clock speed, but for now this code just assumes like the others do
- missing @alevy's idea of EitherBuffer which would make sending buffer slices through the UART simpler

### CODE REVIEW

File: chips/sam4l/src/spi.rs [hunk 1]
```diff
@@ -418,23 +418,17 @@ impl DMAClient for Spi {
...
-
-            let rxbuf = match self.dma_read.as_mut() {
-                Some(dma) => {
-                    let buf = dma.abort_xfer();
-                    dma.disable();
-                    buf
-                }
-                None => None,
-            };
+            let txbuf = self.dma_write.map_or(None, |dma| {
```
Comment (alevy) [inline chips/sam4l/src/spi.rs]: Oooo... much nicer.

File: chips/sam4l/src/usart.rs [hunk 2]
```diff
@@ -1,130 +1,311 @@
...
-    ifr: u32,
-    man: u32,
-    linmr: u32,
-    linir: u32,
-    linbrr: u32,
-    wpmr: u32,
-    wpsr: u32,
-    version: u32,
+struct USARTRegisters {
+    cr: VolatileCell<u32>, // 0x00
```
Comment (alevy) [inline chips/sam4l/src/usart.rs]: Yes please and thank you

File: chips/sam4l/src/usart.rs [hunk 3]
```diff
@@ -145,127 +326,260 @@ impl USART {
...
+        regs.rtor.set(0);
+
+        // enable timeout interrupt
+        regs.idr.set((1 << 8)); // TIMEOUT
+    }
+
+    fn enable_rx_terminator(&self, terminator: u8) {
+        let regs: &mut USARTRegisters = unsafe { mem::transmute(self.registers) };
+        // XXX: implement me
+        panic!("didn't write terminator stuff yet");
```
Comment (alevy) [inline chips/sam4l/src/usart.rs]: Is the plan for this to basically just receive one byte at a time?

File: chips/sam4l/src/usart.rs [hunk 4]
```diff
@@ -1,131 +1,311 @@
...
 
-    pub fn set_dma(&mut self, dma: &'static mut DMAChannel, dma_peripheral: DMAPeripheral) {
-        self.dma.replace(dma);
-        self.dma_peripheral = dma_peripheral;
+    pub fn set_dma(&self, rx_dma: &'static dma::DMAChannel, tx_dma: &'static dma::DMAChannel) {
+        self.rx_dma.replace(rx_dma);
+        self.tx_dma.replace(tx_dma);
     }
 
-    fn set_baud_rate(&self, baud_rate: u32) {
```
Comment (bradjc) [inline chips/sam4l/src/usart.rs]: It might be good to keep the `set_baud_rate()` function for future USART implementations (SPI, etc.).

**Comment by bradjc** [comment]
So what do we want to do with this? Simplify the HIL? Make a UART advanced HIL?

**Comment by alevy** [comment]
@bradjc I think yes and yes.

I can take this over and do that since we'll also need to port for the nrf51

**Comment by bradjc** [comment]
There are definitely some fixes in the signpost branch that also need to get included here, both to console, nrf-serialization, and the usart driver itself.

File: boards/imix/src/io.rs [hunk 5]
```diff
@@ -14,17 +14,20 @@ impl Write for Writer {
...
             uart.enable_tx();
 
         }
+        // XXX: I'd like to get this working the "right" way, but I'm not sure how
         for c in s.bytes() {
             uart.send_byte(c);
+            while !uart.tx_ready() {}
         }
         Ok(())
     }
```
Comment (brghena) [inline boards/imix/src/io.rs]: While I was in this file, I noticed that it has `... fn panic_fmt(...` whereas storm/src/io.rs has `... fn rust_begin_unwind(...`. I know the storm one works. Is the imix one correct, @alevy?

**Comment by brghena** [comment]
Okay, finally back to this PR. I pulled in the updates that we made in the `signpost` branch while working on demo stuff.

To your point about overly complicated UART function, I do not disagree. However, I found the `receive_automatic` function essential for the implementation of nrf-serialization and for GPS stuff used on signpost. (see below for technical reasons).

I propose (and have implemented) them as a separate `UARTAdvanced` trait. On platforms where they are not easy to implement and haven't proven necessary they can be ignored and capsules that only require normal UART functionality (like `console`) will still work.

---
Technical discussion if you're interested

On the signpost controller, the sam4l is running at 16 MHz. nrf-serialization runs at 250000 baud. During testing, I discovered that the amount of time spent to process an incoming byte and send a buffer to the user was long enough that we would drop the next byte in the packet. So reading the length bytes, setting up a callback for the user, and then starting the next receive based on the value of those length bytes took too long to do. The solution was the move to `receive_automatic` and do the parsing of data after an entire buffer had been received.

For GPS, the problem is simpler. The NMEA standard doesn't actually have a fixed length for messages. So if you're in Michigan, your longitude degrees are two digits (83) and the total packet length is N bytes, but in California they are three digits (122) and the total packet length is N+1. The NMEA standard doesn't have a length field, but instead has message terminators (\r\n). `receive_until_terminator` is exactly what this driver needed. Since I had never gotten around to implementing that, `receive_automatic` worked as a good stand-in.

**Comment by alevy** [comment]
I will port to the NRF when I get back from OSDI tomorrow (don't have one on me), then I think we shold be able to merge.

File: kernel/src/hil/uart.rs [hunk 6]
```diff
@@ -1,42 +1,84 @@
...
 #[derive(Copy, Clone)]
-pub enum Parity {
-    Even = 0,
-    Odd = 1,
-    ForceZero = 2,
-    ForceOne = 3,
-    None = 4,
-    Multidrop = 6,
+pub enum StopBits {
+    One = 0,
```
Comment (bradjc) [inline kernel/src/hil/uart.rs]: This seems odd.

File: capsules/src/nrf51822_serialization.rs [hunk 7]
```diff
@@ -156,16 +173,26 @@ impl<'a, U: UART> Driver for Nrf51822Serialization<'a, U> {
                 });
                 result.unwrap_or(-1)
             }
+            9001 => {
```
Comment (bradjc) [inline capsules/src/nrf51822_serialization.rs]: Should change the numbers to something reasonable.

**Pull request merged**